src/dune_engine/action_runner.mli

```diff
@@ -0,0 +1,40 @@
+open Import
+
+module Foreman : sig
+  (** A foreman manages workers connected to it. This module should therefore be
+      used by the build server *)
+
+  type t
+
+  val create : unit -> t
+
+  type worker
+
+  (** Wait for a worker to terminate. You must run this function for every
+      worker *)
+  val await : worker -> unit Fiber.t
+
+  (** [register pid ~name] must be called after a worker process has been
+      started. The worker process should have a pid of [pid] and is named
+      [name]. *)
+  val worker : t -> Pid.t -> name:string -> worker
+
+  (** [exec_action worker action] dispatches [action] to [worker] *)
+  val exec_action :
+    worker -> Action_exec.input -> Action_exec.Exec_result.t Fiber.t
+
+  (** [implement_handler handler] adds the functions needed to manage connected
+      workers to the rpc server. This only needs to be called by the RPC server *)
+  val implement_handler : t -> 'a Dune_rpc_server.Handler.t -> unit
+end
+
+module Worker : sig
+  (** A worker is a runner of action *)
+
+  (** [start ~name ~where] start a runner named [name] connected to server
+      [where]. The server is allowed to dispatch actions to this worker once the
+      worker initializes itself.
+
+      This function returns when the connection to the server terminates. *)
+  val start : name:string -> where:Dune_rpc_private.Where.t -> unit Fiber.t
+end
```

Comment:
The API of this module seems pretty ornate. I would expect something much simpler, like:
```
  type t

  val start : name:string -> t Fiber.t
  val run_action : t -> Action_exec.Input.t -> Action_exec.Result.t Fiber.t
  val stop : t -> unit Fiber.t
```
And, possibly, some `init` function to set things up correctly for the Dune server.

I guess I am still missing something.

Reply:
> I would expect something much simpler,

Did you attach your comment to the wrong line? `Worker` is a module with a singe function. Can't go much simpler than that :)

I see two problems with your proposed API:

1. The PID isn't present anywhere. So this means that the caller must reap it. To me this is means that two different modules are responsible for the "liftetime" of the worker so there's a failure to isolate concerns

2. There's no "await" function or something similar to catch exceptions. Keep in mind that in fiber there's no background exceptions. So where would you catch errors that aren't related to any particular action? In particular, this  is where I currently raise if the worker has been killed by a signal for example.

Reply:
Sorry, my comment referred to the whole `action_runner.mli`.

Reply:
> The PID isn't present anywhere. So this means that the caller must reap it.

You seem to be talking about pretty low-level details. Why should they be exposed via the high-level `Action_runner` API?

> There's no "await" function or something similar to catch exceptions.

OK, if we need to capture errors while starting/executing actions, then how about the following?

```ocaml
module Action_runner : sig
  type t

  (** Start the action runner using a wrapper passed to Dune via the command line. *)
  val start : name:string -> (t, exn) result Fiber.t

  (** Run an action using the action runner [t]. *) 
  val run_action : t -> Action_exec.Input.t -> (Action_exec.Result.t, exn) result Fiber.t

  (** Stop the action runner [t]. *)
  val stop : t -> (unit, exn) result Fiber.t
end
```

Reply:
Basically, I think the foremen itself should be an implementation detail. I don't see why Dune engine needs to know about it, since the only thing it cares about is executing actions via different action runners.

Reply:
> You seem to be talking about pretty low-level details. Why should they be exposed via the high-level Action_runner API?

Yes, that's worth considering. I think we can consider adding a `unit Fiber.t` that must be waited on instead. But note that it would make debugging a bit tougher as we wouldn't have the `pid` <-> `name` association in some places to make the errors better.

> Basically, I think the foremen itself should be an implementation detail. I don't see why Dune engine needs to know about it, since the only thing it cares about is executing actions via different action runners.

Yup, that would be ideal. But note that your proposed API handles "background" errors quite differently. In particular, suppose the action runner disconnects (say it was killed by a signal). Your api only allows receiving the error on the next `run_action` call. While my API will receive the error as soon as `await` is called. I don't know if that matters in this case, but I usually prefer API's that report errors promptly.

Reply:
Oh and I forgot one more detail - the foreman maintains the following invariant:

All connected runners must have unique names. I don't know if we this, but it doesn't seem like a bad thing to have.

Reply:
> But note that it would make debugging a bit tougher as we wouldn't have the `pid` <-> `name` association in some places to make the errors better.

> In particular, suppose the action runner disconnects (say it was killed by a signal).

I think it makes sense for `Action_runner` to provide an API that does not assume that there is a one to one mapping between `Action_runner.t`s and processes and/or connections. Whether it's just one process, or a group of processes perhaps distributed across multiple machines, does not really matter to Dune's engine.

In the first version, I would expect that if `run_action` returns an `Error`, Dune would treat it as if the action failed, i.e. it would continue building other actions that may be using other action runners [1].

> I usually prefer API's that report errors promptly.

I agree that there is some value in prompt error reporting. However, given the grand plans of eventually making action runners remote, wouldn't it make sense to attempt to reconnect in case of network problems, or more generally handle these background errors in the background, instead of promptly failing? 

[1] As a point of reference, when Jenga's forker process is killed, Jenga happily keeps building and reporting one error per action, like `Error communicating to jenga forker: ((rpc_error (Connection_closed ("EOF or connection closed")))`. That is perhaps not ideal but I think it's useful to follow what it's doing in the first version.

Reply:
Okay, I'll hide the pid. I thought about doing it, it's just that it seemed like getting something working was more important than designing something general. I don't think it will be much more work to drop it though.

> I agree that there is some value in prompt error reporting. However, given the grand plans of eventually making action runners remote, wouldn't it make sense to attempt to reconnect in case of network problems, or more generally handle these background errors in the background, instead of promptly failing?

Right, it would make sense. Except that I think that the explicit `await` makes that easier!

 Keep in mind that your proposed API only allows reconnecting on `run_action` - because that is the only point at which we can catch errors from the action runner. While with an explicit `await`, we don't necessarily need to just fail, we can immediately try reconnecting after a disconnect (say with a backoff).

>  As a point of reference, when Jenga's forker process is killed, Jenga happily keeps building and reporting one error per action, like Error communicating to jenga forker: ((rpc_error (Connection_closed ("EOF or connection closed"))). That is perhaps not ideal but I think it's fine to follow what Jenga is doing in the first version.

That seems contrary to the automatic reconnect idea, I'll do that for v1 though.

Reply:
A small comment that is hopefully not off topic:

In async, there's this useful function `don't_wait_for` that allows one to do work in the background. Such as trying to reconnect whenever a disconnect occurs. With fibers we (fortunately?) don't have this luxury and need to bind on every single `'a Fiber.t` that may be encountered. So this forces us to add various `await` or `run` functions whose sole purpose is to collect background errors and do background work (usually with the help of pools). So it's often the case that we end up with less ergonomic api's.

Reply:
Thanks! Just to clarify a bit more: my comments here are driven primarily by the desire to make the v1 API as simple as possible (not by some long-term grand plans). 

> Keep in mind that your proposed API only allows reconnecting on `run_action`

I see, that makes sense, and I think it's just fine. We won't need to reconnect in the foreseeable future. As I said, Jenga's implementation is very simplistic, so the bar is not set very high. Let's not raise it too early.

Reply:
One last distinction in the API:

```
 (** Start the action runner using a wrapper passed to Dune via the command line. *)
  val start : name:string -> (t, exn) result Fiber.t
```

1. I'd rather `action_runner` in the engine not know any details about the command line. As we've discussed a few times, the command line is still ephemeral and even if it was stable, I'd rather everything related to the command line to live in one place (bin/).

2. Note that your API does not allow to concurrently initialize the runners and run build actions. Or rather, that would be something that would need to be implemented on top unlike the current API. Perhaps that's the right way to do it though.

Reply:
Agreed that everything related to the command line should live in one place. That's what I meant by maybe needing some kind of `init` function in the API, so that we can pass the necessary information into the `Action_runner` module at startup.

> Note that your API does not allow to concurrently initialize the runners and run build actions

That seems right, yes. Maybe it's worth acknowledging this limitation somewhere in a comment.

Reply:
> A small comment that is hopefully not off topic:
> 
> In async, there's this useful function `don't_wait_for` that allows one to do work in the background. Such as trying to reconnect whenever a disconnect occurs. With fibers we (fortunately?) don't have this luxury and need to bind on every single `'a Fiber.t` that may be encountered. So this forces us to add various `await` or `run` functions whose sole purpose is to collect background errors and do background work (usually with the help of pools). So it's often the case that we end up with less ergonomic api's.

Oops, I just noticed this comment. Agreed, but structured concurrency comes with some benefits too.